libantworld/agent.h

```diff
@@ -21,67 +35,279 @@ using namespace units::literals;
+        glfwWindowHint(GLFW_RESIZABLE, false);
+
+        GLFWwindow *ptr = glfwCreateWindow(size.width, size.height, "Ant world", nullptr, nullptr);
+        if (!ptr) {
+            glfwTerminate();
+            throw std::runtime_error("Failed to create window");
+        }
+
+        // Wrap in a unique_ptr so we can free it properly when we're done
+        std::unique_ptr<GLFWwindow, std::function<void(GLFWwindow *)>> window(ptr, &glfwDestroyWindow);
+
+        // Make the window's context current
+        glfwMakeContextCurrent(window.get());
+
+        // Initialize GLEW
+        if (glewInit() != GLEW_OK) {
+            throw std::runtime_error("Failed to initialize GLEW");
+        }
+
+        // Enable VSync
+        glfwSwapInterval(1);
+
+        glDebugMessageCallback(handleGLError, nullptr);
+
+        // Set clear colour to match matlab and enable depth test
+        glClearColor(0.0f, 1.0f, 1.0f, 1.0f);
+        glEnable(GL_DEPTH_TEST);
+        glLineWidth(4.0);
+        glPointSize(4.0);
+
+        glEnable(GL_CULL_FACE);
+        glCullFace(GL_BACK);
+
+        glEnable(GL_TEXTURE_2D);
+
+        return window;
+    }
+
 private:
-    Vector3<degree_t> m_Attitude{{ 0_deg, 0_deg, 0_deg }};
-    Vector3<meter_t> m_Position{{ 0_m, 0_m, 0_m }};
+    using TimeType = std::chrono::time_point<std::chrono::high_resolution_clock>;
+
+    Pose3<meter_t, degree_t> m_Pose;
+    meters_per_second_t m_Velocity;
+    radians_per_second_t m_TurnSpeed;
     Renderer &m_Renderer;
     GLFWwindow *m_Window;
+    std::mutex m_PoseMutex;
+    float m_JoystickX = 0.f, m_JoystickY = 0.f;
+
+    Stopwatch m_MoveStopwatch;
+    enum class MoveMode
+    {
+        NotMoving,
+        MovingForward,
+        Turning
+    } m_MoveMode = MoveMode::NotMoving;
+    float m_MoveSpeed;
+
+    void updatePose()
+    {
+        const units::time::second_t elapsed = m_MoveStopwatch.lap();
+
+        using namespace units::math;
+        switch (m_MoveMode) {
+        case MoveMode::MovingForward: {
+            const meter_t dist = m_MoveSpeed * m_Velocity * elapsed;
+            m_Pose.x() += dist * cos(m_Pose.yaw());
+            m_Pose.y() += dist * sin(m_Pose.yaw());
+        } break;
+        case MoveMode::Turning:
+            m_Pose.yaw() -= m_MoveSpeed * m_TurnSpeed * elapsed;
+            while (m_Pose.yaw() > 360_deg) {
+                m_Pose.yaw() -= 360_deg;
+            }
+            while (m_Pose.yaw() < 0_deg) {
+                m_Pose.yaw() += 360_deg;
+            }
+            break;
+        default:
+            break;
+        }
+    }
+
+    static void handleGLFWError(int errorNumber, const char *message)
+    {
+        throw std::runtime_error("GLFW error number: " + std::to_string(errorNumber) + ", message:" + message);
+    }
+
+    static void handleGLError(GLenum, GLenum, GLuint, GLenum, GLsizei, const GLchar *message, const void *)
+    {
+        throw std::runtime_error(message);
+    }
+
+    static TimeType now()
```

Comment:
What's this for? It looks like you're using the stop watch for calculating elapsed time between updates

Reply:
Oops, must've forgotten to take this out